host_core/lib/host_core/actors/actor_module.ex

```diff
@@ -246,91 +319,221 @@ defmodule HostCore.Actors.ActorModule do
-                       inv["content_length"],
-                       Map.get(inv, "msg", <<>>)
-                     )
-                     |> IO.iodata_to_binary()
-                   ) do
-                {:ok, response} ->
-                  Tracer.set_status(:ok, "")
-
-                  {%{
-                     msg: response,
-                     invocation_id: inv["id"],
-                     instance_id: iid,
-                     content_length: byte_size(response)
-                   }
-                   |> chunk_inv_response(), inv}
-
-                {:error, error} ->
-                  Logger.error("Invocation failure: #{error}", invocation_id: inv["id"])
-                  Tracer.set_status(:error, "Invocation failure: #{error}")
-
-                  {%{
-                     msg: <<>>,
-                     error: error,
-                     invocation_id: inv["id"],
-                     instance_id: iid
-                   }, inv}
-              end
+      publish_invocation_result(host_id, lattice_prefix, token.invocation, ir)
+
+      {:reply, {:ok, ir |> Msgpax.pack!() |> IO.iodata_to_binary()}, agent}
+    end
+  end
+
+  defp unpack_body(%{} = token, body) do
+    case Msgpax.unpack(body) do
+      {:ok, inv} ->
+        Tracer.set_attribute("invocation_id", inv["id"])
+
+        %{token | invocation: inv}
+
+      _ ->
+        Tracer.set_status(:error, "Failed to deserialize msgpack invocation")
+
+        %{
+          token
+          | inv_res: %{
+              msg: <<>>,
+              invocation_id: "",
+              error: "Failked to deserialize invocation",
+              instance_id: token.iid
+            }
+        }
+    end
+  end
+
+  defp validate_anti_forgery(%{invocation: nil} = token, _body, _issuers) do
+    %{token | anti_forgery: false}
+  end
+
+  defp validate_anti_forgery(%{invocation: inv} = token, body, issuers) do
+    case HostCore.WasmCloud.Native.validate_antiforgery(
+           body,
+           issuers
+         ) do
+      {:error, msg} ->
+        Logger.error("Invocation failed anti-forgery validation check: #{msg}",
+          invocation_id: inv["id"]
+        )
+
+        Tracer.set_status(:error, "Anti-forgery check failed #{msg}")
+
+        %{
+          token
+          | anti_forgery: false,
+            inv_res: %{
+              msg: <<>>,
+              invocation_id: token.invocation["id"],
+              error: "Anti-forgery check failed: #{msg}",
+              instance_id: token.iid
+            }
+        }
+
+      _ ->
+        %{token | anti_forgery: true}
+    end
+  end
+
+  defp validate_invocation_source_target(%{anti_forgery: false} = token, _agent) do
+    %{token | source_target: false}
+  end
+
+  defp validate_invocation_source_target(
+         %{
+           anti_forgery: true,
+           invocation: %{
+             "origin" => %{
+               "link_name" => nil,
+               "contract_id" => nil
+             }
+           }
```

Comment:
this is how it used to be as well ;) The reason is because the invocation is decoded from the wire and that gives us string keys, whereas the `invocation` field on state is an atom because we wrote it. I've tried changing this and it failed miserably enough that I decided to leave it for later.